File: desktop/extension-host/tests/render_element.ts
import "wasi";
import { Console } from "as-wasi";
import { JSON, JSONEncoder } from "assemblyscript-json";

export function render_element(): void {
  let props = RenderElementProps.fromJSON(
    <JSON.Obj>JSON.parse(Console.readAll())
  );

  const encoder = new JSONEncoder()

  encoder.setString("", `<p ${props.attributes.toHTML()}>foobar</p>`);

  Console.log(encoder.toString());
}

class RenderElementProps {
  element: Element;
  attributes: RenderElementAttributes;

  constructor(element: Element, attributes: RenderElementAttributes) {
    this.element = element;
    this.attributes = attributes;
  }

  static fromJSON(json: JSON.Obj): RenderElementProps {
    return new RenderElementProps(
      Element.fromJSON(json.getObj("element")!),
      RenderElementAttributes.fromJSON(json.getObj("attributes")!)
    );
  }
}

class Element {
  type: string;
  children: JSON.Value[];

  constructor(type: string, children: JSON.Value[]) {
    this.type = type;
    this.children = children;
  }

  static fromJSON(json: JSON.Obj): Element {
    return new Element(
      json.getString("type")!.valueOf(),
      json.getArr("children")!.valueOf()
    );
  }
}

class RenderElementAttributes {
  dataSlateNode: string;
  dataSlateInline: boolean;
  dataSlateVoid: boolean;
  dir: string | null;
  ref: JSON.Value;

  constructor(
    dataSlateNode: string,
    dataSlateInline: boolean,
    dataSlateVoid: boolean,
    dir: string | null,
    ref: JSON.Value
  ) {
    this.dataSlateNode = dataSlateNode;
    this.dataSlateInline = dataSlateInline;
    this.dataSlateVoid = dataSlateVoid;
    this.dir = dir;
    this.ref = ref;
  }

  static fromJSON(json: JSON.Obj): RenderElementAttributes {
    let dir: string | null = null
    if (json.getString('dir') && json.getString('dir')!.isString) {
      dir = json.getString('dir')!.valueOf();
    }

    return new RenderElementAttributes(
      json.getString("data-slate-node")!.valueOf(),
      json.getBool("data-slate-inline")!.valueOf(),
      json.getBool("data-slate-void")!.valueOf(),
      dir,
      json.get("ref")!
    );
  }

  toHTML(): string {
    let str = `data-slate-node="${this.dataSlateNode.toString()}" data-slate-inline="${this.dataSlateInline.toString()}" data-slate-void="${this.dataSlateVoid.toString()}" ref="${this.ref.toString()}"`;
    if (this.dir) {
      str += ` dir="${this.dir!}"`
    }
    return str;
  }
}

File: desktop/extension-host/tests/render_element.rs
use cap_std::fs::Dir;
use common::{Element, RenderElementAttributes, RenderElementProps};
use extension_host::{hooks::render_element::HostExt, FsModuleLoader, Host};

#[tokio::test]
pub async fn main() -> anyhow::Result<()> {
  tracing_subscriber::fmt::init();
  let dir = Dir::open_ambient_dir("./tests", cap_std::ambient_authority())?;
  let mut host = Host::new(FsModuleLoader::new(dir));

  host.load_extension("./render_element.wasm").await?;

  let props = RenderElementProps {
    children: Vec::new(),
    element: Element {
      r#type: "paragraph".to_string(),
      children: Vec::new(),
    },
    attributes: RenderElementAttributes {
      data_slate_node: "element".to_string(),
      data_slate_inline: true,
      data_slate_void: true,
      dir: Some("rtl".to_string()),
      r#ref: serde_json::Value::Null,
    },
  };

  let html = host.render_element(&props).await?;

  assert_eq!(html, "<p data-slate-node=\"element\" data-slate-inline=\"true\" data-slate-void=\"true\" ref=\"null\" dir=\"rtl\">foobar</p>");

  Ok(())
}
